Build failure due to implicit-function-declaration and int-conversion

**Describe the bug**
Trying to build cx_Freeze on Fedora Linux Rawhide results in several errors:
```
  gcc -fno-strict-overflow -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -fcf-protection -fexceptions -fcf-protection -fexceptions -fcf-protection -fexceptions -O3 -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -mtls-dialect=gnu2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -fPIC -I/usr/include/python3.13 -c source/bases/console.c -o build/temp.linux-x86_64-cpython-313/source/bases/console.o
  In file included from source/bases/console.c:43:
  source/bases/common.c: In function ‘InitializePython’:
  source/bases/common.c:289:5: warning: ‘Py_NoSiteFlag’ is deprecated [-Wdeprecated-declarations]
    289 |     Py_NoSiteFlag = 1;
        |     ^~~~~~~~~~~~~
  In file included from /usr/include/python3.13/Python.h:72,
                   from source/bases/console.c:7:
  /usr/include/python3.13/cpython/pydebug.h:14:37: note: declared here
     14 | Py_DEPRECATED(3.12) PyAPI_DATA(int) Py_NoSiteFlag;
        |                                     ^~~~~~~~~~~~~
  source/bases/common.c:290:5: warning: ‘Py_FrozenFlag’ is deprecated [-Wdeprecated-declarations]
    290 |     Py_FrozenFlag = 1;
        |     ^~~~~~~~~~~~~
  /usr/include/python3.13/cpython/pydebug.h:16:37: note: declared here
     16 | Py_DEPRECATED(3.12) PyAPI_DATA(int) Py_FrozenFlag;
        |                                     ^~~~~~~~~~~~~
  source/bases/common.c:291:5: warning: ‘Py_IgnoreEnvironmentFlag’ is deprecated [-Wdeprecated-declarations]
    291 |     Py_IgnoreEnvironmentFlag = 1;
        |     ^~~~~~~~~~~~~~~~~~~~~~~~
  /usr/include/python3.13/cpython/pydebug.h:17:37: note: declared here
     17 | Py_DEPRECATED(3.12) PyAPI_DATA(int) Py_IgnoreEnvironmentFlag;
        |                                     ^~~~~~~~~~~~~~~~~~~~~~~~
  source/bases/common.c:292:5: warning: ‘Py_SetProgramName’ is deprecated [-Wdeprecated-declarations]
    292 |     Py_SetProgramName(wexecutable);
        |     ^~~~~~~~~~~~~~~~~
  In file included from /usr/include/python3.13/Python.h:119:
  /usr/include/python3.13/pylifecycle.h:37:38: note: declared here
     37 | Py_DEPRECATED(3.11) PyAPI_FUNC(void) Py_SetProgramName(const wchar_t *);
        |                                      ^~~~~~~~~~~~~~~~~
  source/bases/common.c:293:5: error: implicit declaration of function ‘Py_SetPath’; did you mean ‘Py_GetPath’? [-Wimplicit-function-declaration]
    293 |     Py_SetPath(wpath);
        |     ^~~~~~~~~~
        |     Py_GetPath
  source/bases/common.c:295:5: warning: ‘PySys_SetArgvEx’ is deprecated [-Wdeprecated-declarations]
    295 |     PySys_SetArgvEx(argc, argv, 0);
        |     ^~~~~~~~~~~~~~~
  In file included from /usr/include/python3.13/Python.h:121:
  /usr/include/python3.13/sysmodule.h:11:38: note: declared here
     11 | Py_DEPRECATED(3.11) PyAPI_FUNC(void) PySys_SetArgvEx(int, wchar_t **, int);
        |                                      ^~~~~~~~~~~~~~~
  source/bases/console.c: In function ‘main’:
  source/bases/console.c:75:14: error: implicit declaration of function ‘_PyMem_RawStrdup’; did you mean ‘PyMem_RawFree’? [-Wimplicit-function-declaration]
     75 |     oldloc = _PyMem_RawStrdup(setlocale(LC_ALL, NULL));
        |              ^~~~~~~~~~~~~~~~
        |              PyMem_RawFree
  source/bases/console.c:75:12: error: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
     75 |     oldloc = _PyMem_RawStrdup(setlocale(LC_ALL, NULL));
        |            ^
  error: command '/usr/bin/gcc' failed with exit code 1
  error: subprocess-exited-with-error
  
  × Building wheel for cx_Freeze (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /usr/bin/python3 /usr/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py build_wheel /builddir/build/BUILD/cx_freeze-7.2.1-build/cx_Freeze-7.2.1/.pyproject-builddir/tmp37lgia8a
  cwd: /builddir/build/BUILD/cx_freeze-7.2.1-build/cx_Freeze-7.2.1
  Building wheel for cx_Freeze (pyproject.toml): finished with status 'error'
  ERROR: Failed building wheel for cx_Freeze
Failed to build cx_Freeze
ERROR: Failed to build one or more wheels
```

**Desktop (please complete the following information):**
 - Platform information (e.g. Ubuntu Linux 22.04): Fedora Linux Rawhide (42)
 - OS architecture (e.g. amd64): x86_64
 - cx_Freeze version [e.g. 6.11]: 7.2.1
 - Python version [e.g. 3.10]: 3.13

**Additional context**
`-Wimplicit-function-declaration` and `-Wint-conversion` are threated as errors in default Fedora compiler flags.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -75,7 +75,12 @@ jobs:
         shell: bash
     env:
       UV_SYSTEM_PYTHON: true
+      UV_NO_PROGRESS: true
     steps:
+
+      - name: Install dependencies (Linux)
+        if: runner.os == 'Linux'
+        run: sudo apt-get install -qy alien fakeroot rpm
 
       - uses: actions/checkout@v4
 
@@ -98,10 +103,6 @@ jobs:
           pattern: cx-freeze-whl-${{ matrix.os }}*
           path: wheelhouse
 
-      - name: Install dependencies (Linux)
-        if: runner.os == 'Linux'
-        run: sudo apt-get install -qy alien fakeroot rpm
-
       - name: Install dependencies to test
         run: |
           uv pip install -r requirements.txt -r tests/requirements.txt
@@ -111,6 +112,69 @@ jobs:
         env:
           COVERAGE_FILE: ".coverage.${{ matrix.python-version }}.${{ matrix.os }}"
         run: pytest -nauto --cov="cx_Freeze"
+
+      - name: Upload coverage reports
+        uses: actions/upload-artifact@v4
+        with:
+          name: cov-${{ matrix.python-version }}.${{ matrix.os }}
+          path: .coverage.*
+          include-hidden-files: true
+
+  test_free_threaded:
+    needs:
+      - build_wheel
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-14]
+        python-version: ['3.13t']
+      fail-fast: false
+    defaults:
+      run:
+        shell: bash
+    env:
+      UV_NO_PROGRESS: true
+    steps:
+
+      - name: Install dependencies (Linux)
+        if: runner.os == 'Linux'
+        run: sudo apt-get install -qy alien fakeroot rpm
+
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ inputs.branch }}
+          repository: marcelotduarte/cx_Freeze
+
+      - uses: astral-sh/setup-uv@v5
+        with:
+          enable-cache: true
+          cache-dependency-glob: |
+            requirements.txt
+            tests/requirements.txt
+          python-version: ${{ matrix.python-version }}
+
+      - name: Env
+        run: env | sort
+
+      - name: Sysconfig
+        run: uv run --no-project -m sysconfig
+
+      - name: Download the wheel
+        uses: actions/download-artifact@v4
+        with:
+          merge-multiple: true
+          pattern: cx-freeze-whl-${{ matrix.os }}*
+          path: wheelhouse
+
+      - name: Install dependencies to test
+        run: |
+          uv sync --no-install-project --extra tests
+          uv pip install cx_Freeze --no-index --no-deps -f wheelhouse --reinstall
+
+      - name: Generate coverage report
+        env:
+          COVERAGE_FILE: ".coverage.${{ matrix.python-version }}.${{ matrix.os }}"
+        run: uv run --no-project pytest -nauto --cov="cx_Freeze"
 
       - name: Upload coverage reports
         uses: actions/upload-artifact@v4

--- a/ci/build-wheel.sh
+++ b/ci/build-wheel.sh
@@ -22,8 +22,9 @@ PY_PLATFORM=$($PYTHON -c "import sysconfig; print(sysconfig.get_platform(), end=
 PY_VERSION=$($PYTHON -c "import sysconfig; print(sysconfig.get_python_version(), end='')")
 PY_VERSION_FULL=$($PYTHON -c "import sysconfig; print(sysconfig.get_config_var('py_version'), end='')")
 PY_VERSION_NODOT=$($PYTHON -c "import sysconfig; print(sysconfig.get_config_var('py_version_nodot'), end='')")
+PY_ABI_THREAD=$($PYTHON -c "import sysconfig; print(sysconfig.get_config_var('abi_thread') or '', end='')")
 
-PYTHON_TAG=cp$PY_VERSION_NODOT
+PYTHON_TAG=cp$PY_VERSION_NODOT$PY_ABI_THREAD
 if [[ $PY_PLATFORM == linux* ]]; then
     PLATFORM_TAG=many$(echo $PY_PLATFORM | sed 's/\-/_/')
     PLATFORM_TAG_MASK="$(echo $PLATFORM_TAG | sed 's/_/*_/')"
@@ -81,8 +82,8 @@ _bump_my_version () {
 
 _cibuildwheel () {
     local args=$*
-    # Use python >= 3.11
-    local py_version=$(_vergte $PY_VERSION 3.11)
+    # Use python >= 3.11 (python 3.13t is supported)
+    local py_version=$(_vergte $PY_VERSION 3.11)$PY_ABI_THREAD
     # Do not export UV_* to avoid conflict with uv in cibuildwheel macOS/Windows
     unset UV_SYSTEM_PYTHON
     uvx -p $py_version cibuildwheel $args
@@ -101,7 +102,7 @@ echo "::endgroup::"
 mkdir -p wheelhouse >/dev/null
 if [[ $PY_PLATFORM == linux* ]]; then
     echo "::group::Build sdist"
-    uv build -p $PY_VERSION --no-build-isolation --sdist -o wheelhouse
+    uv build -p $PY_VERSION$PY_ABI_THREAD --sdist -o wheelhouse
     echo "::endgroup::"
 fi
 echo "::group::Build wheel(s)"

--- a/cx_Freeze/_compat.py
+++ b/cx_Freeze/_compat.py
@@ -7,6 +7,7 @@ import sysconfig
 from pathlib import Path
 
 __all__ = [
+    "ABI_THREAD",
     "BUILD_EXE_DIR",
     "EXE_SUFFIX",
     "EXT_SUFFIX",
@@ -21,8 +22,9 @@ __all__ = [
 
 PLATFORM = sysconfig.get_platform()
 PYTHON_VERSION = sysconfig.get_python_version()
+ABI_THREAD = sysconfig.get_config_var("abi_thread") or ""
 
-BUILD_EXE_DIR = Path(f"build/exe.{PLATFORM}-{PYTHON_VERSION}")
+BUILD_EXE_DIR = Path(f"build/exe.{PLATFORM}-{PYTHON_VERSION}{ABI_THREAD}")
 EXE_SUFFIX = sysconfig.get_config_var("EXE")
 EXT_SUFFIX = sysconfig.get_config_var("EXT_SUFFIX")
 

--- a/cx_Freeze/executable.py
+++ b/cx_Freeze/executable.py
@@ -11,6 +11,7 @@ from sysconfig import get_config_var
 from typing import TYPE_CHECKING
 
 from cx_Freeze._compat import (
+    ABI_THREAD,
     EXE_SUFFIX,
     IS_MACOS,
     IS_MINGW,
@@ -74,12 +75,17 @@ class Executable:
 
     @base.setter
     def base(self, name: str | Path | None) -> None:
-        # The default base is the legacy console, except for
+        # The default base is the legacy console, except for Python 3.13t and
         # Python 3.13 on macOS, that supports only the new console
-        if IS_MACOS and sys.version_info[:2] >= (3, 13):
-            name = name or "console"
-        else:
+        version = sys.version_info[:2]
+        if (
+            version <= (3, 13)
+            and ABI_THREAD == ""
+            and not (IS_MACOS and version == (3, 13))
+        ):
             name = name or "console_legacy"
+        else:
+            name = name or "console"
         # silently ignore gui and service on non-windows systems
         if not (IS_WINDOWS or IS_MINGW) and name in ("gui", "service"):
             name = "console"

--- a/cx_Freeze/freezer.py
+++ b/cx_Freeze/freezer.py
@@ -22,11 +22,13 @@ from zipfile import ZIP_DEFLATED, ZIP_STORED, PyZipFile, ZipFile, ZipInfo
 from setuptools import Distribution
 
 from cx_Freeze._compat import (
+    ABI_THREAD,
     BUILD_EXE_DIR,
     IS_CONDA,
     IS_MACOS,
     IS_MINGW,
     IS_WINDOWS,
+    PYTHON_VERSION,
 )
 from cx_Freeze.common import get_resource_file_path, process_path_specs
 from cx_Freeze.exception import FileError, OptionError
@@ -1035,9 +1037,10 @@ class WinFreezer(Freezer, PEParser):
             # MSYS2 python returns a static library.
             names = [name.replace(".dll.a", ".dll")]
         else:
+            py_version = f"{PYTHON_VERSION}{ABI_THREAD}"
             names = [
                 f"python{sys.version_info[0]}.dll",
-                f"python{sys.version_info[0]}{sys.version_info[1]}.dll",
+                f"python{py_version.replace('.','')}.dll",
             ]
         python_shared_libs: list[Path] = []
         for name in names:
@@ -1113,7 +1116,7 @@ class DarwinFreezer(Freezer, Parser):
     def _default_bin_includes(self) -> list[str]:
         python_shared_libs: list[Path] = []
         # Check for distributed "cx_Freeze/bases/lib/Python"
-        name = "Python"
+        name = f"Python{ABI_THREAD.upper()}"
         for bin_path in self._default_bin_path_includes():
             fullname = Path(bin_path, name).resolve()
             if fullname.is_file():

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -61,7 +61,7 @@ dynamic = ["version"]
 
 [project.optional-dependencies]
 dev = [
-    "bump-my-version==0.29.0",
+    "bump-my-version==0.29.0 ;python_version < '3.13'",
     "cibuildwheel==2.22.0",
     "pre-commit==4.0.1",           # python_version >= 3.9
 ]
@@ -164,9 +164,11 @@ optional_value = "final"
 before-build = "uv pip install -r requirements.txt"
 build-frontend = "build[uv]"
 build-verbosity = 1
+enable = ["cpython-freethreading"]
 skip = [
     "cp3{9,10,13}-musllinux_*",
     "cp3{9,10,13}-manylinux_ppc64le",
+    "cp313t-win*",
 ]
 
 [tool.cibuildwheel.linux]

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -1,3 +1,3 @@
-bump-my-version==0.29.0
+bump-my-version==0.29.0 ;python_version < '3.13'
 cibuildwheel==2.22.0
 pre-commit==4.0.1

--- a/setup.py
+++ b/setup.py
@@ -97,7 +97,8 @@ class BuildBases(setuptools.command.build_ext.build_ext):
             library_dirs.append(get_config_var("LIBPL"))
             if not ENABLE_SHARED or IS_CONDA:
                 library_dirs.append(get_config_var("LIBDIR"))
-            libraries.append(f"python{get_python_version()}")
+            abi_thread = get_config_var("abi_thread") or ""
+            libraries.append(f"python{get_python_version()}{abi_thread}")
             if get_config_var("LIBS"):
                 extra_args.extend(get_config_var("LIBS").split())
             if get_config_var("LIBM"):
@@ -275,38 +276,48 @@ def get_extensions() -> list[Extension]:
         os.environ.get("CI", "") != "true"
         or os.environ.get("CIBUILDWHEEL", "0") != "1"
     )
+    abi_thread = get_config_var("abi_thread") or ""
+    version = sys.version_info[:2]
     extensions = [
         Extension(
             "cx_Freeze.bases.console",
             ["source/bases/console.c", "source/bases/_common.c"],
             optional=optional,
-        ),
-        Extension(
-            "cx_Freeze.bases.console_legacy",
-            ["source/legacy/console.c"],
-            depends=["source/legacy/common.c"],
-            optional=optional
-            or (sys.version_info[:2] >= (3, 13) and IS_MACOS),
-        ),
+        )
     ]
-
-    if IS_MINGW or IS_WINDOWS:
+    if (
+        version <= (3, 13)
+        and abi_thread == ""
+        and not (IS_MACOS and version == (3, 13))
+    ):
         extensions += [
             Extension(
-                "cx_Freeze.bases.Win32GUI",
-                ["source/legacy/Win32GUI.c"],
+                "cx_Freeze.bases.console_legacy",
+                ["source/legacy/console.c"],
                 depends=["source/legacy/common.c"],
-                libraries=["user32"],
                 optional=optional,
-            ),
-            Extension(
-                "cx_Freeze.bases.Win32Service",
-                ["source/legacy/Win32Service.c"],
-                depends=["source/legacy/common.c"],
-                extra_link_args=["/DELAYLOAD:cx_Logging"],
-                libraries=["advapi32"],
-                optional=optional,
-            ),
+            )
+        ]
+    if IS_MINGW or IS_WINDOWS:
+        if version <= (3, 13) and abi_thread == "":
+            extensions += [
+                Extension(
+                    "cx_Freeze.bases.Win32GUI",
+                    ["source/legacy/Win32GUI.c"],
+                    depends=["source/legacy/common.c"],
+                    libraries=["user32"],
+                    optional=optional,
+                ),
+                Extension(
+                    "cx_Freeze.bases.Win32Service",
+                    ["source/legacy/Win32Service.c"],
+                    depends=["source/legacy/common.c"],
+                    extra_link_args=["/DELAYLOAD:cx_Logging"],
+                    libraries=["advapi32"],
+                    optional=optional,
+                ),
+            ]
+        extensions += [
             Extension(
                 "cx_Freeze.bases.gui",
                 ["source/bases/Win32GUI.c", "source/bases/_common.c"],

--- a/tests/test_executables.py
+++ b/tests/test_executables.py
@@ -12,6 +12,7 @@ from setuptools import Distribution
 
 from cx_Freeze import Executable
 from cx_Freeze._compat import (
+    ABI_THREAD,
     BUILD_EXE_DIR,
     EXE_SUFFIX,
     IS_MACOS,
@@ -241,14 +242,18 @@ TEST_VALID_PARAMETERS = [
         ("icon.ico", "icon.icns", "icon.png", "icon.svg"),
     ),
 ]
-if IS_MACOS and sys.version_info[:2] >= (3, 13):
-    TEST_VALID_PARAMETERS += [
-        ("base", None, "console-"),
-    ]
-else:
+if (
+    sys.version_info[:2] <= (3, 13)
+    and ABI_THREAD == ""
+    and not (IS_MACOS and sys.version_info[:2] == (3, 13))
+):
     TEST_VALID_PARAMETERS += [
         ("base", None, "console_legacy-"),
         ("base", "console_legacy", "console_legacy-"),
+    ]
+else:
+    TEST_VALID_PARAMETERS += [
+        ("base", None, "console-"),
     ]
 if IS_WINDOWS or IS_MINGW:
     TEST_VALID_PARAMETERS += [

--- a/tests/test_freezer.py
+++ b/tests/test_freezer.py
@@ -12,6 +12,7 @@ from generate_samples import create_package, run_command
 
 from cx_Freeze import Freezer
 from cx_Freeze._compat import (
+    ABI_THREAD,
     BUILD_EXE_DIR,
     EXE_SUFFIX,
     IS_CONDA,
@@ -99,19 +100,20 @@ def test_freezer_default_bin_includes(tmp_path: Path, monkeypatch) -> None:
     monkeypatch.chdir(tmp_path)
 
     freezer = Freezer(executables=["hello.py"])
+    py_version = f"{PYTHON_VERSION}{ABI_THREAD}"
     if IS_MINGW:
-        expected = f"libpython{PYTHON_VERSION}.dll"
+        expected = f"libpython{py_version}.dll"
     elif IS_WINDOWS:
-        expected = f"python{PYTHON_VERSION.replace('.','')}.dll"
+        expected = f"python{py_version.replace('.','')}.dll"
     elif IS_CONDA:  # macOS or Linux
         if IS_MACOS:
-            expected = f"libpython{PYTHON_VERSION}.dylib"
+            expected = f"libpython{py_version}.dylib"
         else:
-            expected = f"libpython{PYTHON_VERSION}.so*"
+            expected = f"libpython{py_version}.so*"
     elif IS_MACOS:
-        expected = "Python"
+        expected = f"Python{ABI_THREAD.upper()}"
     elif ENABLE_SHARED:  # Linux
-        expected = f"libpython{PYTHON_VERSION}.so*"
+        expected = f"libpython{py_version}.so*"
     else:
         assert freezer.default_bin_includes == []
         return

